warehouse/forklift/legacy.py

```diff
@@ -758,7 +766,8 @@ def _is_duplicate_file(db_session, filename, hashes):
 
     if file_ is not None:
         return (
-            file_.filename == filename
+            # This has the effect of canonicalizing the project name and version
+            _parse_filename(file_.filename) == _parse_filename(filename)
```

Comment:
Is the intention here to ensure that I can't upload `foo_bar-0.1.0.tar.gz` if `foo.bar-0.1.0.tar.gz` already exists?

If so, I don't think this quite handles it. In the query above we filter on if the filenames OR hashes collide.

In the test for this change, we always collide hashes... and I'm pretty sure it would fail to catch the "collision" marked above.

Reply:
It’s also not quite easy to check for non-canonicalized file names from the database here, at least not without some schema shakeup. This is sort of a “best effort” to detect the user uploading the same file; duplicates are still possible, but they wouldn’t be _that_ problematic and it’s not worth it to eliminate them.